### PR TITLE
Fix polars

### PR DESCRIPTION
**Pull request by ritchie46**
Hey, I am from the Polars project and was surprised Polars was added to ClickBench. When I looked at the implementation I saw that Polars was incorrectly used (#268). and sometimes even python functions were called where we have a native API for the operations.

This ensures that 

- Polars' queries can use the Lazy API and thus the optimizer and the query engine. 
- The statistics are computed in single query (similar to SQL). That ensures we can parallelize
- Use Polars expressions and not `map_elements` with python udfs

Resolves: https://github.com/ClickHouse/ClickBench/issues/268

### CODE REVIEW

File: polars/query.py [hunk 1]
```diff
@@ -22,48 +22,48 @@
...
         "SELECT COUNT(*) FROM hits WHERE AdvEngineID <> 0;",
         lambda x: x[x["AdvEngineID"] != 0].count(),
-        lambda x: x.filter(pl.col("AdvEngineID") != 0).height,
+        lambda x: x.select(pl.col("AdvEngineID").filter(pl.col("AdvEngineID") != 0).count()).collect().height,
     ),
     (
         "Q2",
         "SELECT SUM(AdvEngineID), COUNT(*), AVG(ResolutionWidth) FROM hits;",
         lambda x: (x["AdvEngineID"].sum(), x.shape[0], x["ResolutionWidth"].mean()),
-        lambda x: (x["AdvEngineID"].sum(), x.height, x["ResolutionWidth"].mean()),
```
Comment (ritchie46) [inline polars/query.py:42]: These are 2 queries, where in SQL it is in a single query.

File: polars/query.py [hunk 2]
```diff
@@ -22,48 +22,48 @@
...
+        lambda x: x.select(pl.col("advengineid").sum(), pl.len(), pl.col("advengineid").mean()).collect().rows()[0],
     ),
     (
         "Q3",
         "SELECT AVG(UserID) FROM hits;",
         lambda x: x["UserID"].mean(),
-        lambda x: x["UserID"].mean(),
```
Comment (ritchie46) [inline polars/query.py:48]: This is not lazy suitable. Forces a single engine. It is meant for notebooks.

File: polars/query.py [hunk 3]
```diff
@@ -318,15 +318,13 @@
         .group_by("CounterID")  # GROUP BY CounterID
         .agg(
             [
-                pl.col("URL")
-                .map_elements(lambda y: len(y), return_dtype=pl.Int64)
```
Comment (ritchie46) [inline polars/query.py:322]: This was wrong. Didn't compute the `AVG` and it uses a python function, where we have native expressions for this.

File: polars/query.py [hunk 4]
```diff
@@ -352,18 +350,14 @@
             .group_by("k")
             .agg(
                 [
-                    pl.col("Referer").map_elements(
-                        lambda y: len(y), return_dtype=pl.Int64
```
Comment (ritchie46) [inline polars/query.py:356]: This was wrong. Didn't compute the `AVG` and it uses a python function, where we have native expressions for this.

**Comment by ritchie46** [comment]
Yes, current code uses pandas to read parquet. I don't know why?

The error you see is a pandas installation error. Shall I remove pandas from that script?

**Comment by rschu1ze** [comment]
> Yes! and if you can run the benchmark c6a.4xlarge, 500gb gp2 (like in datafusion) it would be great!

This relates to #264. I actually tried to run polars/benchmark.sh on my corporate c6a.4xlarge machine (using the version in this PR, now that [the setup is fixed](https://github.com/ClickHouse/ClickBench/pull/270#issuecomment-2499660384)) ... and it froze my machine so badly that it did not come back in 2.5 hours after which I killed the instance forcefully. Perhaps someone else has more luck.

**Comment by ritchie46** [comment]
> and it froze my machine so badly that I was unable to connect via SSH or ping it for 2.5 hours,

Yes, this version loads all data in memory which doesn't fit that machine. This was already the case, and I didn't adapt that in the PR.

I can also follow up with a PR where we load data from parquet. That doesn't materialize all data and is much less memory intensive.

**Comment by rschu1ze** [comment]
> I can also follow up with a PR where we load data from parquet. That doesn't materialize all data and is much less memory intensive.

That would be nice. Actually, we should try to do changes to the benchmark scripts / queries and the results themselves in the same PR. Otherwise, one of them will be outdated. So I guess my preference is if we could update all polars variants within this PR (and if it helps, I can setup myself a `c6a.metal` box which is _way_ more powerful than `c6a.4xlarge` and try to redo the measurements).

**Comment by rschu1ze** [comment]
@ritchie46 Side question: Like to sign above CLA declaration as well? I don't care too much personally but the lawyers might. Thanks.

**Comment by ritchie46** [comment]
> @ritchie46 Side question: Like to sign above CLA declaration as well? I don't care too much personally but the lawyers might. Thanks.

Signed. :+1:

**Comment by ritchie46** [comment]
@rschu1ze I have added parquet as a source as well. Made it so that we create 2 output jsons, indicating the source used.

Note that running from parquet has drastically reduced memory usage. The DataFrame case is just very intensive as we are forced to load all data in memory.

**Comment by rschu1ze** [comment]
@ritchie46 Thanks ... I'll re-run measurements in the coming days!

**Comment by ritchie46** [comment]
Great, thanks!

**Comment by rschu1ze** [comment]
I both dataframes and parquet versions on c6a.4xlarge again but the system hang. Didn't check deeper why, instead I repeated the measurements on a c6a.metal machine. Pushed the measurements into this PR.

**Comment by alexey-milovidov** [comment]
@rschu1ze, the server's freeze means OOM. Please run it with `prlimit` as here:
https://github.com/ClickHouse/ClickBench/pull/267/files#diff-6ad3e2e988a332b62ba8cc145121a402e3372f9414784d7c9a93e7b6967e8be6R11

Some queries will be shown as failed, but the results are still good to publish (and then they can be improved).

**Comment by rschu1ze** [comment]
Done: https://github.com/ClickHouse/ClickBench/pull/279

**Pull request merged**